src/include/miopen/graphapi/graphapi_tensor.hpp

```diff
@@ -0,0 +1,190 @@
+
+class Tensor
+{
+public:
+    Tensor() noexcept         = default;
+    Tensor(const Tensor&)     = default;
+    Tensor(Tensor&&) noexcept = default;
+    Tensor& operator=(const Tensor&) = default;
+    Tensor& operator=(Tensor&&) noexcept = default;
+    Tensor(miopenDataType_t dataType,
+           const std::vector<int64_t>& dimensions,
+           const std::vector<int64_t>& strides,
+           int64_t id,
+           bool isVirtual)
+        : mDimensions(dimensions),
+          mStrides(strides),
+          mId(id),
+          mDataType(dataType),
+          mVirtual(isVirtual)
+    {
+    }
+    Tensor(miopenDataType_t dataType,
+           std::vector<int64_t>&& dimensions,
+           std::vector<int64_t>&& strides,
+           int64_t id,
+           bool isVirtual) noexcept
+        : mDimensions(std::move(dimensions)),
+          mStrides(std::move(strides)),
+          mId(id),
+          mDataType(dataType),
+          mVirtual(isVirtual)
+    {
+    }
+
+    miopenDataType_t getDataType() const noexcept { return mDataType; }
+    const std::vector<int64_t>& getDimensions() const noexcept { return mDimensions; }
+    const std::vector<int64_t>& getStrides() const noexcept { return mStrides; }
+    int64_t getId() const noexcept { return mId; }
+    bool getVirtual() const noexcept { return mVirtual; }
+
+    Tensor& setDataType(miopenDataType_t dataType) noexcept
+    {
+        mDataType = dataType;
+        return *this;
+    }
+    Tensor& setDimensions(const std::vector<int64_t>& dimensions)
+    {
+        mDimensions = dimensions;
+        return *this;
+    }
+    Tensor& setDimensions(std::vector<int64_t>&& dimensions) noexcept
+    {
+        mDimensions = std::move(dimensions);
+        return *this;
+    }
+    Tensor& setStrides(const std::vector<int64_t>& strides)
+    {
+        mStrides = strides;
+        return *this;
+    }
+    Tensor& setStrides(std::vector<int64_t>&& strides) noexcept
+    {
+        mStrides = std::move(strides);
+        return *this;
+    }
+    Tensor& setId(int64_t id) noexcept
+    {
+        mId = id;
+        return *this;
+    }
+    Tensor& setVirtual(bool isVirtual) noexcept
+    {
+        mVirtual = isVirtual;
+        return *this;
+    }
+
+private:
+    std::vector<int64_t> mDimensions;
+    std::vector<int64_t> mStrides;
+    int64_t mId                = 0;
+    miopenDataType_t mDataType = miopenFloat;
+    bool mVirtual              = false;
+};
+
+class TensorBuilder
+{
+public:
+    TensorBuilder& setDataType(miopenDataType_t dataType) &;
+    TensorBuilder& setDim(const std::vector<int64_t>& dimensions) &;
+    TensorBuilder& setDim(std::vector<int64_t>&& dimensions) &;
+    TensorBuilder& setStride(const std::vector<int64_t>& strides) &;
+    TensorBuilder& setStride(std::vector<int64_t>&& strides) &;
+    TensorBuilder& setId(int64_t id) &;
+    TensorBuilder& setVirtual(bool isVirtual) &;
+
+    TensorBuilder&& setDataType(miopenDataType_t dataType) &&
```

Comment:
setters should not return an `r-value` reference or be marked as `r-value` methods.

Reply:
this is for preserving r-value references in sequences like 
`auto result = Builder().setSomething().setSomethingElse().setAnotherSomething().build();`

Reply:
cool.